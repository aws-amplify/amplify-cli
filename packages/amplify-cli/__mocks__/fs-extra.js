const fsExtra = jest.genMockFromModule('fs-extra');

function copySync() {
  return {};
}

fsExtra.copySync = copySync;

module.exports = fsExtra;
