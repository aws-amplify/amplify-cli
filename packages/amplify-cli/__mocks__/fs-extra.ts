const fs: any = jest.genMockFromModule('fs-extra');

fs.readdirSync = jest.fn().mockReturnValue(['file1', 'file2']);
fs.readFileSync = jest.fn().mockReturnValue({});
fs.statSync = jest.fn().mockReturnValue({ isDirectory: () => {} });
fs.unlinkSync = jest.fn().mockReturnValue(null);
fs.copySync = jest.fn().mockReturnValue({});

module.exports = fs;
