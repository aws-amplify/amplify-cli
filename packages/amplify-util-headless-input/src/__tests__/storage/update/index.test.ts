import * as fs from 'fs-extra';
import * as path from 'path';
import { validateUpdateStorageRequest } from '../../../index';

const topLevelMissingProperty: ExpectedFunction = (missingProperty: string) =>
  `Data did not validate against the supplied schema. Underlying errors were [{\"keyword\":\"required\",\"dataPath\":\"\",\"schemaPath\":\"#/required\",\"params\":{\"missingProperty\":\"${missingProperty}\"},\"message\":\"should have required property '${missingProperty}'\"}]`;
const serviceModificationMissingProperty: ExpectedFunction = (missingProperty: string) =>
  `Data did not validate against the supplied schema. Underlying errors were [{\"keyword\":\"required\",\"dataPath\":\".serviceModification\",\"schemaPath\":\"#/required\",\"params\":{\"missingProperty\":\"${missingProperty}\"},\"message\":\"should have required property '${missingProperty}'\"}]`;

type ExpectedFunction = (missingProperty: string) => string;

const updateStorageAssetsPath = path.resolve(path.join(__dirname, '..', '..', 'assets', 'storage', 'update'));

describe('validate update storage request', () => {
  it('returns valid payload', async () => {
    const rawRequest = fs.readFileSync(path.join(updateStorageAssetsPath, 'validUpdateStorageRequest.json'), 'utf8');
    const result = await validateUpdateStorageRequest(rawRequest);
    expect(result).toMatchSnapshot();
  });
});

describe('rejects promise when invalid payload', () => {
  it('rejects garbage', async () => {
    const resultPromise = validateUpdateStorageRequest('garbage');
    await expect(resultPromise).rejects.toBeTruthy();
  });

  test.each([
    [
      'version',
      'invalidRequest.missing.version.json',
      (missingProperty: string) => `data does not have a top level \"${missingProperty}\" field`,
    ],
    [
      'version',
      'invalidRequest.string.version.json',
      (missingProperty: string) => `data does not have a top level \"${missingProperty}\" field`,
    ],
    ['version', 'invalidRequest.invalid.version.json', (missingProperty: string) => `No schema found for ${missingProperty} 9999`],
    ['serviceModification', 'invalidRequest.missing.serviceModification.json', topLevelMissingProperty],
    ['serviceName', 'invalidRequest.missing.serviceName.json', serviceModificationMissingProperty],
    ['resourceName', 'invalidRequest.missing.resourceName.json', serviceModificationMissingProperty],
    ['permissions', 'invalidRequest.missing.permissions.json', serviceModificationMissingProperty],
  ])(
    'with a rejected promise when the payload is missing %s',
    async (missingProperty: string, fileName: string, expectedFn: ExpectedFunction) => {
      const rawRequest = fs.readFileSync(path.join(updateStorageAssetsPath, fileName), 'utf8');

      await expect(validateUpdateStorageRequest(rawRequest)).rejects.toThrow(expectedFn(missingProperty));
    },
  );
});
