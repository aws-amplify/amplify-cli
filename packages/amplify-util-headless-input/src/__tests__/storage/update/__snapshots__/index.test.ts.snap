// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`validate update storage request returns valid payload 1`] = `
Object {
  "serviceModification": Object {
    "bucketName": "something",
    "lambdaTrigger": Object {
      "mode": "existing",
      "name": "triggerName",
    },
    "permissions": Object {
      "auth": Array [
        "READ",
        "CREATE_AND_UPDATE",
      ],
      "groups": Object {
        "myUserGroup1": Array [
          "CREATE_AND_UPDATE",
          "READ",
          "DELETE",
        ],
        "myUserGroup2": Array [
          "CREATE_AND_UPDATE",
          "READ",
        ],
        "myUserGroup3": Array [
          "CREATE_AND_UPDATE",
        ],
      },
      "guest": Array [
        "READ",
      ],
    },
    "resourceName": "something",
    "serviceName": "S3",
  },
  "version": 1,
}
`;
